### ISSUE TITLE
Copy, cut and paste with context menu

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab_app#206]
### Problem

First, thanks for the awesome work on reviving this! I tested on Ubuntu and it works!

There is the `Shift + Right Click for Browser Menu` entry, but `Shift + Right Click` does not do anything for me. This means that some actions (copy, paste, cut, copy link address) are not accessible now.

![Screenshot from 2021-09-15 17-31-31](https://user-images.githubusercontent.com/5832902/133476306-b9bfaf1c-830d-4c98-977a-8cc13ee314af.png)

### Proposed Solution

- Work with upstream JupterLab to add full self-contained functionality of copy, paste, and cut (which are already available in Text Editor but not in Notebook) and maybe additional actions (copy link) too
- Disable the `Shift + Right Click for Browser Menu` entry in JupyterLab App as it is misleading (this should be possible with the menu schema files).

**Comment by krassowski** [comment]
I explored four approaches:
- A) using `webContents` API (e.g. `webContents.copy`) for both interacting with clipboard and with the page, and communicating via remotes
- B) using `Electron.clipboard` directly for interacting with clipboard, Web APIs like [`Selection`](https://developer.mozilla.org/en-US/docs/Web/API/Selection) and DOM API for interacting with the page, and communicating via remotes
- C) using [`Clipboard API`](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API) on client side only, no communication with server
- D) adding a fallback context menu with Electron-native implementation of commonly needed actions

The code for A and B would rougly look as follows:

<details>
<summary>Code</summary>

In `desktop-extension`

```ts
app.commands.addCommand(CommandIDs.copySelection, {
    label: 'Copy Text',
    execute: () => {
        const selection = window.getSelection();
        asyncRemoteRenderer.runRemoteMethod(IAppRemoteInterface.writeToClipboard, selection.toString()).catch(console.warn);
    },
    isVisible: () => {
        const selection = window.getSelection();
        return selection.toString() !== '';
    },
});
```

and server side (in `main/app`):

```ts
asyncRemoteMain.registerRemoteMethod(IAppRemoteInterface.writeToClipboard,
    (data: string): Promise<void> => {
        // approach A:
        // clipboard.writeText(data);
        // approach B:
        this._window.webContents.copy();
        return Promise.resolve();
    });
```

</details>

Problems with approaches A and B:
   - `isVisible()` gets called too frequently leading to performance degradation (rendering it unusable);
   - handling of CodeEditor in Notebook cells is difficult because the custom context menu takes focus away from the editor hence the text is no longer selected; this necessitates writing custom code as for the [`Copy` command of `File Editor` in the core JupyterLab](https://github.com/jupyterlab/jupyterlab/blob/8a8c3752564f37493d4eb6b4c59008027fa83880/packages/fileeditor-extension/src/commands.ts#L807-L819) but also aware of multiple cells (which would be a slight adjustment)

Problems with A:
- in order to handle `CodeEditor`s issue requires some clever focus management/closing the JupyterLab context menu first, ensuring focus is back on selected element and only then executing the action. This is doable but will require a major API change upstream to implement it cleanly. This might be error-prone.

Problems with B:
- in order to handle `CodeEditor`s issue requires getting the selected text directly from the editor
- requires re-implementing a bit of logic to handle rich-MIME copying; this is already somewhat implemented in JupyterLab core and we would be duplicating it here if using approach B
- pasting is difficult (especially into `contentEditable`/`CodeMirror` editor) because we need to manage DOM nodes/events manually; this is not an issue with approach A.
- "Copy link URL" and "Copy Image" might not be trivial to implement well (but still feasible!)

Problems with C:
- it needs to be implemented directly in JupyterLab
   - JupyterLab postponed implementation waiting for wider browser support for Clipboard API
   - it might need to be implemented as an opt-in because some browsers do not allow to read from clipboard in web context
- actions such as "Copy link URL" and "Copy Image" will need to be implemented (note: we already have "Copy Output to Clipboard" implemented in https://github.com/jupyterlab/jupyterlab/pull/10282)

Problems with D:
- we end up with two menus again, this is detrimental to UX (though at least consistent with the web version)

I think that long term we should pursue approach C by implementing relevant code in JupyterLab core (4.0 or 4.1?). In the meantime we can adopt approach D.

**Comment by notluquis** [comment]
Going through this one: both halves of what you proposed in 2021 have happened, so it may be closeable.

**The interim approach (D) shipped.** `LabView._addFallbackContextMenu` gives cut/copy/paste/select all in editable fields, copy on a selection, **Copy link URL** on links and **Copy image** on images. Of your original list only "Save image" is missing, which you had marked as a maybe. The misleading `Shift + Right Click for Browser Menu` entry is gone from the codebase; the only trace left is a comment.

**The long-term approach (C) landed upstream, and we ship it.** jupyterlab/jupyterlab#18030 merged on 2026-05-19 in milestone 4.6.0, and `env_installer/jlab_server.yaml` pins jupyterlab 4.6.2. Unpacking that exact wheel, the commands are there: `notebook:copy-selected-text`, `notebook:cut-selected-text`, `notebook:paste-text` and `fileeditor:copy` / `cut` / `paste`, with `navigator.clipboard` behind them. So text copy/paste comes from JupyterLab's own menu in the version the app installs, which is the condition the comment in our code names for dropping the fallback.

**How the two menus interact**, since I had to check rather than assume. In the same 4.6.2 bundle the handler reads:

```js
evtContextMenu(e) {
  this._contextMenuEvent = e;
  if (e.shiftKey || suppressContextMenu(e.target)) { return; }
  if (this.contextMenu.open(e)) { ...; e.preventDefault(); e.stopPropagation(); }
}
```

So the Shift bypass is JupyterLab's own, not something Chromium does for us, and `preventDefault` only happens when JupyterLab actually opened a menu. On the Electron side I measured that a page calling `preventDefault` on `contextmenu` stops the `context-menu` event from firing at all, shift held or not. Put together: where JupyterLab shows a menu, ours cannot appear; where it has nothing for that target, ours appears on a plain right click.

What is left is narrower than the issue title:

- **Copy link URL** and **Copy image** have no upstream equivalent. Searching the 4.6.2 bundle for those command labels returns nothing, and the only related thing upstream is "Copy Output to Clipboard". Removing the fallback outright would lose them.
- "Save image" was never implemented.
- Discoverability, rather than capability: the fallback is reachable, but only where JupyterLab declines the event or the user knows to hold Shift.

Do you remember anything else you were waiting on here, or is this one to close with the remaining pieces tracked separately if anyone wants them? What I have not done is sit in front of the app and right click on a dozen surfaces; the above is from the shipped bundle and from an Electron probe.

Checked with Claude Code.

**Comment by krassowski** [comment]
I think we can close this and address issues like context menu and search UI in JupyterLab itself. A lot has improved here, not only in notebook:
- https://github.com/jupyterlab/jupyterlab/pull/13535
- https://github.com/jupyterlab/jupyterlab/pull/19195
- https://github.com/jupyterlab/jupyterlab/pull/17691
- https://github.com/jupyterlab/jupyterlab/pull/18030 (already mentioned)

Just today I opened two issues on the same theme:
- https://github.com/jupyterlab/jupyterlab/issues/19213
- https://github.com/jupyterlab/jupyterlab/issues/19214

I would find a lot of value in having a generic "Copy Image" in JupyterLab menu whenever Context menu is open over a user-generated menu.
